File: plugin/vtabs/glyphs.lua
local util = require "vtabs.util"

local M = {}

---Chrome glyphs, their ASCII fallbacks and the group they substitute with.
local CHROME = {
  chamfer_top = { "▙", " ", group = "block" },
  chamfer_bottom = { "▛", " ", group = "block" },
  scroll = { "▐", "|", group = "block" },
  active = { "▎", "|", group = "bar" },
  unseen = { "•", "*", group = "marks" },
  focus = { "›", ">", group = "marks" },
  ellipsis = { "…", "...", group = "marks" },
  meta_sep = { "  ", "  ", group = "marks" },
  toggle_left = { "«", "<", group = "toggle" },
  toggle_right = { "»", ">", group = "toggle" },
  frame_tl = { "╭", "+", group = "ghost_frame" },
  frame_tr = { "╮", "+", group = "ghost_frame" },
  frame_bl = { "╰", "+", group = "ghost_frame" },
  frame_br = { "╯", "+", group = "ghost_frame" },
  frame_dash = { "╌", "-", group = "ghost_frame" },
  frame_dash_v = { "╎", "|", group = "ghost_frame" },
  rule = { "─", "-", group = "marks" },
}

---East Asian Ambiguous: width flips with unicode_version / treat_east_asian_ambiguous_width_as_wide.
local AMBIGUOUS = {
  [0x258e] = true,
  [0x2022] = true,
  [0x2026] = true,
  [0x00b7] = true,
  [0x00ab] = true,
  [0x00bb] = true,
  [0x256d] = true,
  [0x256e] = true,
  [0x256f] = true,
  [0x2570] = true,
  [0x2500] = true,
}

local function ambiguous(s)
  if type(s) ~= "string" or utf8.len(s) ~= 1 then
    return false
  end
  return AMBIGUOUS[utf8.codepoint(s)] == true
end

local function group_members(group)
  local keys = {}
  for key, spec in pairs(CHROME) do
    if spec.group == group then
      keys[#keys + 1] = key
    end
  end
  table.sort(keys)
  return keys
end

---Resolves the chrome glyphs for one window against its effective config.
---@param base table icons.resolve output; process icons pass through untouched
---@param effective table|nil window:effective_config(), or nil for the width backstop only
function M.resolve(base, effective)
  base = base or {}
  effective = effective or {}
  local out = {}
  for key, value in pairs(base) do
    out[key] = value
  end
  for key, spec in pairs(CHROME) do
    out[key] = base[key] or spec[1]
  end
  out.corners = "chamfer"

  local by_width = false
  local function fall_back(key)
    local spec = CHROME[key]
    if spec and out[key] ~= spec[2] then
      out[key] = spec[2]
    end
  end
  local function fall_back_group(group)
    for _, key in ipairs(group_members(group)) do
      fall_back(key)
    end
  end

  if effective.custom_block_glyphs == false then
    out.corners = "square"
    fall_back_group "block"
    fall_back_group "bar"
  end

  -- only this flag selects ambiguous width; unicode_version alone never does
  local wide_ambiguous = effective.treat_east_asian_ambiguous_width_as_wide == true
  if wide_ambiguous then
    for key in pairs(CHROME) do
      if ambiguous(out[key]) then
        fall_back(key)
      end
    end
    fall_back_group "ghost_frame"
  end

  for key in pairs(CHROME) do
    if util.width(out[key]) ~= 1 then
      fall_back(key)
      by_width = true
    end
  end
  for _, key in ipairs(group_members "ghost_frame") do
    if out[key] ~= CHROME[key][1] then
      fall_back_group "ghost_frame"
      break
    end
  end
  for _, key in ipairs { "close", "pinned", "private", "new_tab", "settings", "search" } do
    if out[key] and util.width(out[key]) ~= 1 then
      out[key] = ({ close = "x", pinned = "*", private = "~", new_tab = "+", settings = "*", search = "/" })[key]
      by_width = true
    end
  end
  if out.chamfer_top == CHROME.chamfer_top[2] then
    out.corners = "square"
  end

  if by_width then
    util.warn_once("glyph-width", "some glyphs are not one cell wide; using ASCII")
  end
  return out
end

return M

File: backend/crates/vtabs-view/src/glyphs.rs
use std::collections::BTreeMap;

use crate::text::width;

/// name, glyph, ASCII fallback, substitution group
const CHROME: &[(&str, &str, &str, &str)] = &[
    ("chamfer_top", "▙", " ", "block"),
    ("chamfer_bottom", "▛", " ", "block"),
    ("scroll", "▐", "|", "block"),
    ("active", "▎", "|", "bar"),
    ("unseen", "•", "*", "marks"),
    ("focus", "›", ">", "marks"),
    ("ellipsis", "…", "...", "marks"),
    ("meta_sep", "  ", "  ", "marks"),
    ("toggle_left", "❮", "<", "toggle"),
    ("toggle_right", "❯", ">", "toggle"),
    ("frame_tl", "╭", "+", "ghost_frame"),
    ("frame_tr", "╮", "+", "ghost_frame"),
    ("frame_bl", "╰", "+", "ghost_frame"),
    ("frame_br", "╯", "+", "ghost_frame"),
    ("frame_dash", "╌", "-", "ghost_frame"),
    ("frame_dash_v", "╎", "|", "ghost_frame"),
    ("rule", "─", "-", "marks"),
    ("hint_up", "↑", "^", "hints"),
    ("hint_down", "↓", "v", "hints"),
    ("hint_left", "←", "<", "hints"),
    ("hint_right", "→", ">", "hints"),
];

/// East Asian Ambiguous: width flips with unicode_version / treat_east_asian_ambiguous_width_as_wide.
const AMBIGUOUS: &[char] = &[
    '\u{258e}', '\u{2022}', '\u{2026}', '\u{00b7}', '\u{256d}', '\u{256e}', '\u{256f}', '\u{2570}',
    '\u{2500}', '\u{2190}', '\u{2191}', '\u{2192}', '\u{2193}',
];

/// Icon keys with a hard ASCII backstop when their configured glyph is not one cell wide.
const ICON_BACKSTOP: &[(&str, &str)] = &[
    ("close", "x"),
    ("pinned", "*"),
    ("private", "~"),
    ("new_tab", "+"),
    ("strip_new_tab", "+"),
    ("settings", "*"),
    ("search", "/"),
];

fn chrome(key: &str) -> Option<(&'static str, &'static str, &'static str)> {
    CHROME
        .iter()
        .find(|(name, ..)| *name == key)
        .map(|&(_, glyph, fallback, group)| (glyph, fallback, group))
}

fn ambiguous(s: &str) -> bool {
    let mut chars = s.chars();
    match (chars.next(), chars.next()) {
        (Some(c), None) => AMBIGUOUS.contains(&c),
        _ => false,
    }
}

pub struct Resolved {
    pub glyphs: BTreeMap<String, String>,
    pub corners: &'static str,
    /// True when a glyph had to fall back over width; Lua warns once on it.
    pub narrowed: bool,
}

/// Resolves the chrome glyphs against the window's effective config; `base` is icons::resolve
/// output, whose process icons pass through untouched.
pub fn resolve(
    base: &BTreeMap<String, String>,
    custom_block_glyphs: bool,
    wide_ambiguous: bool,
) -> Resolved {
    let mut out = base.clone();
    for &(key, glyph, ..) in CHROME {
        out.entry(key.to_string())
            .or_insert_with(|| glyph.to_string());
    }
    let mut corners = "chamfer";
    let mut narrowed = false;

    let fall_back = |out: &mut BTreeMap<String, String>, key: &str| {
        if let Some((_, fallback, _)) = chrome(key)
            && out[key] != fallback
        {
            out.insert(key.to_string(), fallback.to_string());
        }
    };
    let fall_back_group = |out: &mut BTreeMap<String, String>, group: &str| {
        for &(key, _, fallback, g) in CHROME {
            if g == group && out[key] != fallback {
                out.insert(key.to_string(), fallback.to_string());
            }
        }
    };

    if !custom_block_glyphs {
        corners = "square";
        fall_back_group(&mut out, "block");
        fall_back_group(&mut out, "bar");
    }

    // only this flag selects ambiguous width; unicode_version alone never does
    if wide_ambiguous {
        for &(key, ..) in CHROME {
            if ambiguous(&out[key]) {
                fall_back(&mut out, key);
            }
        }
        fall_back_group(&mut out, "ghost_frame");
    }

    for &(key, ..) in CHROME {
        if width(&out[key]) != 1 {
            fall_back(&mut out, key);
            narrowed = true;
        }
    }
    let ghost_intact = CHROME
        .iter()
        .filter(|(.., g)| *g == "ghost_frame")
        .all(|&(key, glyph, ..)| out[key] == glyph);
    if !ghost_intact {
        fall_back_group(&mut out, "ghost_frame");
    }
    for &(key, backstop) in ICON_BACKSTOP {
        if out.get(key).is_some_and(|g| width(g) != 1) {
            out.insert(key.to_string(), backstop.to_string());
            narrowed = true;
        }
    }
    if out["chamfer_top"] == chrome("chamfer_top").unwrap().1 {
        corners = "square";
    }

    Resolved {
        glyphs: out,
        corners,
        narrowed,
    }
}
